src/main/scala/org/constellation/checkpoint/CheckpointBlockValidator.scala

```diff
@@ -111,47 +95,83 @@ class CheckpointBlockValidator[F[_]: Sync](
   snapshotService: SnapshotService[F],
   checkpointParentService: CheckpointParentService[F],
   transactionValidator: TransactionValidator[F],
+  blacklistAddressValidator: BlacklistAddressValidator[F],
+  doubleSpendValidator: DoubleSpendValidator[F],
   dao: DAO
 ) {
 
   import CheckpointBlockValidator._
 
-  val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+  private val logger = Slf4jLogger.getLogger[F]
 
   def simpleValidation(cb: CheckpointBlock): F[ValidationResult[CheckpointBlock]] =
     for {
-      validation <- validateCheckpointBlock(cb)
-      _ <- if (validation.isValid) dao.metrics.incrementMetricAsync("checkpointValidationSuccess")
-      else
-        dao.metrics
-          .incrementMetricAsync(Metrics.checkpointValidationFailure)
-          .flatTap(_ => logger.warn(s"Checkpoint block with baseHash: ${cb.baseHash} is invalid $validation"))
-    } yield validation
-
-  def validateTransactionIntegrity(t: Transaction): F[ValidationResult[Transaction]] =
+      _ <- logger.debug(s"Start Validation for CheckpointBlock : ${cb.baseHash}")
+      validationResult <- validateCheckpointBlock(cb)
+      _ <- logger.debug(s"Validation finish for CheckpointBlock : ${cb.baseHash}")
+
+      _ <- updateMetrics(validationResult, cb.baseHash)
+    } yield validationResult
+
+  private def validateCheckpointBlock(cb: CheckpointBlock): F[ValidationResult[CheckpointBlock]] = {
+    val preTreeResult =
+      for {
+        _ <- logger.debug(s"Static Validation for CheckpointBlock : ${cb.baseHash}")
+        staticValidation <- Sync[F].delay(
+          validateEmptySignatures(cb.signatures).product(validateSignatures(cb.signatures, cb.baseHash))
+        )
+
+        _ <- logger.debug(s"Transactions Validation for CheckpointBlock : ${cb.baseHash}")
+        transactionsValidation <- validateTransactions(cb.transactions)
+
+        _ <- logger.debug(s"DuplicatedTransactions Validation for CheckpointBlock : ${cb.baseHash}")
+        duplicatedTransactions <- Sync[F].delay(validateDuplicatedTransactions(cb.transactions))
+
+        _ <- logger.debug(s"Balance Validation Validation for CheckpointBlock : ${cb.baseHash}")
+        balanceValidation <- validateSourceAddressBalances(cb.transactions)
+
+        _ <- logger.debug(s"DoubleSpend Validation for CheckpointBlock : ${cb.baseHash}")
+        doubleSpendValidator <- doubleSpendValidator.validate(cb.transactions.toList)
+
+        _ <- logger.debug(s"BlacklistAddress Validation for CheckpointBlock : ${cb.baseHash}")
+        blacklistAddressValidator <- blacklistAddressValidator.validate(cb.transactions.toList)
+      } yield
+        staticValidation
+          .product(transactionsValidation)
+          .product(duplicatedTransactions)
+          .product(balanceValidation)
+          .product(blacklistAddressValidator)
+          .product(doubleSpendValidator)
```

Comment:
One thing to note is that if we mark a checkpoint as invalid and only blacklist the address locally, we  lose track of the evidence that a double spend occurred which we will need to keep so that other nodes can update their blacklist. Perhaps we could make `doubleSpendValidator` and `blacklistAddressValidator` a Unit operation and then broadcast the checkpoint block as usual. We will need to modify the `CheckpointAcceptanceService` to call `doubleSpendValidator` and `blacklistAddressValidator` as well. See my comment below: